Source/Particles/ParticleBoundaryBuffer.H

```diff
@@ -0,0 +1,42 @@
+/* Copyright 2021 Andrew Myers
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#ifndef PARTICLEBOUNDARYBUFFER_H_
+#define PARTICLEBOUNDARYBUFFER_H_
+
+#include "Particles/ParticleBuffer.H"
+#include "Particles/MultiParticleContainer_fwd.H"
+
+#include <vector>
+
+/**
+ *  This stores particles that have left / been absorbed by domain and embedded boundaries.
+ */
+class ParticleBoundaryBuffer
+{
+public:
+    ParticleBoundaryBuffer ();
+
+    int nSpecies() const {return m_species_names.size();}
+    std::vector<std::string> GetSpeciesNames() const { return m_species_names; }
+
+    void gatherParticles (MultiParticleContainer& mypc);
+
+    void printNumParticles () const;
+
+    int getNumParticlesInContainer(std::string species_name, int boundary);
```

Comment:
This should take its argument as `const std::string &species_name`, since it doesn't need a local, modifiable copy.